presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java

```diff
@@ -46,15 +53,31 @@
         private final PlanNodeId planNodeId;
         private final Step step;
         private final List<AccumulatorFactory> accumulatorFactories;
+        private final List<Type> sourceTypes;
```

Comment:
Ah... good point. That explains a lot ;)

But still I think it is somehow doable, but it would not be so easy as I thought at first. We could model some entity which would be  consuming AggregationOperator input and would be responsible for providing input for each aggregation. In case of aggregation with NO ORDER BY it could simply pass its input. In case of ORDER BY it could sort appropriately for each aggregation.

The best would be and I hope it is possible when such abstraction could be then used in `HashAggregationOperator` as well.